### DISCUSSION TITLE
Efficiently removing multiple nodes satisfying custom predicate from rtree

### DISCUSSION THREAD

**Discussion by vnair81** [boostorg/geometry#1266] [Q&A] [answered]
I need to remove multiple nodes from rtree which satisfy a custom predicate for which I am currently doing the following:

1. Run a query with the custom predicate and accumulate the node values
2. Sequentially remove each node by its value

This looks inefficient as each rtree node must be visited once for step 1 and then for each identified node, it must again be found in the rtree by the remove method in step 2.

Is there a more efficient method which I could employ?

**Comment by awulkiew** [comment, marked as answer]
Yes, you're correct. This is inefficient. Currently there is no other way.

**Comment by vnair81** [comment]
Another problem I encountered with the current `remove` is that if two nodes have the same value, it will remove either of the two. To remove both I can pass the same value twice to `remove`. However if I want to remove only one but not the other, there is no way for me to specify which of the two I want removed. It will remove the first one that it finds.

In my usage, I add a rtree node for each node in my application's DOM. When one of the DOM node changes, I have to update it in rtree which I am attempting by `remove` followed by `insert`. But because of this arbitrary removal, sometimes the other (unchanged) node gets removed.

Would implementing something similar to `std::remove_if` to accommodate predicates be feasible in the current implementation?

**Reply by awulkiew** [reply]
You could pass your own `EqualTo` (see the [docs](https://beta.boost.org/doc/libs/1_82_0/libs/geometry/doc/html/geometry/reference/spatial_indexes/boost__geometry__index__rtree.html#geometry.reference.spatial_indexes.boost__geometry__index__rtree.synopsis)). You could also store `pair`/`tuple` of e.g. Box and some id allowing the default `EqualTo` to distinguish between objects. Would this work for you?

**Reply by vnair81** [reply]
Thanks @awulkiew.
Default `EqualTo` worked with `tuple`.
